File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/user/UserRelations.java
package top.likoslupus.cellulosesz.api.user;

import java.util.Set;
import java.util.UUID;

import static java.util.Objects.requireNonNull;

public record UserRelations(Set<UUID> ignored) {

    public UserRelations {
        ignored = Set.copyOf(requireNonNull(ignored, "ignored"));
    }

    public static UserRelations defaults() {
        return new UserRelations(Set.of());
    }

    public UserRelations withIgnored(Set<UUID> value) {
        return new UserRelations(value);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/user/UserRelations.kt
package top.likoslupus.cellulosesz.api.user

import top.likoslupus.cellulosesz.api.util.toImmutableSet
import java.util.*

public class UserRelations(
    ignored: Set<UUID>
) {

    @get:JvmName("ignored")
    public val ignored: Set<UUID> = ignored.toImmutableSet()

    public fun withIgnored(value: Set<UUID>): UserRelations =
        UserRelations(value.toImmutableSet())

    override fun equals(other: Any?): Boolean {
        return when {
            this === other -> true
            other !is UserRelations -> false
            else -> ignored == other.ignored
        }
    }

    override fun hashCode(): Int = ignored.hashCode()

    override fun toString(): String = "UserRelations[ignored=$ignored]"

    public companion object {

        @JvmStatic
        public fun defaults(): UserRelations = UserRelations(emptySet())

    }

}
